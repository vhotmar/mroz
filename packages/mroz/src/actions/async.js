import { createAction } from "redux-actions";

export const request = createAction("@@mroz/ASYNC_REQUEST", id => ({
  id
}));

export const success = createAction("@@mroz/ASYNC_SUCCESS", (id, data) => ({
  id,
  data
}));

export const fail = createAction("@@mroz/ASYNC_FAIL", (id, error) => ({
  id,
  error
}));
