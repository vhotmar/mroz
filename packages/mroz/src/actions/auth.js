import { createAction } from "redux-actions";

export const login = createAction("@@mroz/LOGIN", (token, profile) => ({
  token,
  profile
}));

export const logout = createAction("@@mroz/LOGOUT");
