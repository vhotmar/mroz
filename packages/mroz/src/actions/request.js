import { createAction } from "redux-actions";

export const request = createAction(
  "@@mroz/REQUEST_REQUEST",
  (id, key, { isPolling = false } = {}) => ({
    id,
    key,
    isPolling
  })
);

export const success = createAction(
  "@@mroz/REQUEST_SUCCESS",
  (id, key, data) => ({
    id,
    key,
    data,
    time: Date.now()
  })
);

export const fail = createAction("@@mroz/REQUEST_FAIL", (id, key, error) => ({
  id,
  key,
  error
}));

export const cancel = createAction("@@mrot/REQUEST_CANCEL", id => ({
  id
}));
