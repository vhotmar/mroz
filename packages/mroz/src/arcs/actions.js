import { compose } from "ramda";
import { Observable } from "rxjs";
import { addSuccess, addFail, addRequest } from "../epics/async";

const identity = x => x;

export default function actions(request, success, fail) {
  return compose(
    request ? addRequest(config => Observable.of(request(config))) : identity,
    success
      ? addSuccess((data, config) => Observable.of(success(data, config)))
      : identity,
    fail
      ? addFail((error, config) => Observable.of(fail(error, config)))
      : identity
  );
}
