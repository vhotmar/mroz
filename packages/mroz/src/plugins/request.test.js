import { request as requestActions } from "../actions";
import plugin, { reducer, ajax } from "./request";

describe("request plugin", () => {
  const constantDate = new Date("2017-06-13T04:41:20");
  let _now;

  beforeAll(() => {
    _now = Date.now;

    /*eslint no-global-assign:off*/
    Date.now = () => Number(constantDate);
  });

  afterAll(() => {
    Date.now = _now;
  });

  describe("plugin", () => {
    it("should have correct properties", () => {
      const p = plugin()({});

      expect(p).toHaveProperty("meta.dependencies.ajax");
      expect(p).toHaveProperty("reducers.request");
    });

    it("should contain correct reducer", () => {
      const p = plugin()({});

      expect(
        p.reducers.request(undefined, requestActions.request("id1", "key1"))
      ).toMatchSnapshot();
    });
  });

  describe("reducer", () => {
    it("should handle request action", () => {
      expect(
        reducer(undefined, requestActions.request("id1", "key1"))
      ).toMatchSnapshot();
    });

    it("should handle success after request", () => {
      const state = [
        requestActions.request("id1", "key1"),
        requestActions.success("id1", "key1", { data: "foo" })
      ].reduce((p, i) => reducer(p, i), undefined);

      expect(state).toMatchSnapshot();
    });

    it("should handle cancel after request", () => {
      const state = [
        requestActions.request("id1", "key1"),
        requestActions.cancel("key1")
      ].reduce((p, i) => reducer(p, i), undefined);

      expect(state).toMatchSnapshot();
    });

    it("should handle fail after request", () => {
      const state = [
        requestActions.request("id1", "key1"),
        requestActions.fail("id1", "key1", { error: "foo" })
      ].reduce((p, i) => reducer(p, i), undefined);

      expect(state).toMatchSnapshot();
    });

    it("should not reset error after new request", () => {
      const state = [
        requestActions.request("id1", "key1"),
        requestActions.fail("id1", "key1", { error: "foo" }),
        requestActions.request("id1", "key1")
      ].reduce((p, i) => reducer(p, i), undefined);

      expect(state).toMatchSnapshot();
    });

    it("should reset error after new request and success", () => {
      const state = [
        requestActions.request("id1", "key1"),
        requestActions.fail("id1", "key1", { error: "foo" }),
        requestActions.request("id1", "key1"),
        requestActions.success("id1", "key1", { data: "foo" })
      ].reduce((p, i) => reducer(p, i), undefined);

      expect(state).toMatchSnapshot();
    });
  });
});
