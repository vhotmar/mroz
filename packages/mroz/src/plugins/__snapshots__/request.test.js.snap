// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`request plugin plugin should contain correct reducer 1`] = `
Object {
  "byKey": Object {
    "key1": Object {
      "id": "id1",
      "isPolling": false,
      "lastUpdated": undefined,
      "queryCount": 1,
    },
  },
}
`;

exports[`request plugin reducer should handle cancel after request 1`] = `
Object {
  "key1": Object {
    "id": "id1",
    "isPolling": false,
    "lastUpdated": undefined,
    "queryCount": 1,
  },
}
`;

exports[`request plugin reducer should handle fail after request 1`] = `
Object {
  "key1": Object {
    "id": "id1",
    "isPolling": false,
    "lastUpdated": undefined,
    "queryCount": 1,
  },
}
`;

exports[`request plugin reducer should handle request action 1`] = `
Object {
  "key1": Object {
    "id": "id1",
    "isPolling": false,
    "lastUpdated": undefined,
    "queryCount": 1,
  },
}
`;

exports[`request plugin reducer should handle success after request 1`] = `
Object {
  "key1": Object {
    "id": "id1",
    "isPolling": false,
    "lastUpdated": 1497328880000,
    "queryCount": 1,
  },
}
`;

exports[`request plugin reducer should not reset error after new request 1`] = `
Object {
  "key1": Object {
    "id": "id1",
    "isPolling": false,
    "lastUpdated": undefined,
    "queryCount": 2,
  },
}
`;

exports[`request plugin reducer should reset error after new request and success 1`] = `
Object {
  "key1": Object {
    "id": "id1",
    "isPolling": false,
    "lastUpdated": 1497328880000,
    "queryCount": 2,
  },
}
`;
